### PR TITLE
fix(datetime): keyboard navigation of time picker columns

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Users cannot select individual increments of time or the am/pm value, when using only their keyboard in an `ion-datetime`. Users can only use the `space` key, to jump increments of 5. The `ion-popover` has opinionated keyboard event handlers that cancels the default browser behavior.

Issue Number: #24070

## What is the new behavior?

Users can utilize the up/down keys to navigate through the picker columns for hour, minutes and am/pm within an `ion-datetime`. Users can use the `home` and `end` keys to jump to the start or end of the list.

Introduces `cancelBrowserKeyboardEvents` option on `ion-popover` that allows implementations to disable the default behavior of canceling the browser events. This allows the picker column to support the browser behavior of scroll snapping with keyboard navigation.

<video src="https://user-images.githubusercontent.com/13732623/142899828-852106e0-662d-4787-8518-431ce4d09cf5.mp4"></video>

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by liamdebeasi — COMMENTED** [review]
Were you able to figure out why the keyboard was not working in a popover? Scroll containers with scroll snapping have native support for selecting items with the arrow keys, so this functionality should be built into the browsers/we should not need to rebuild it ourselves.

**Comment by sean-perkins** [comment]
> Were you able to figure out why the keyboard was not working in a popover? Scroll containers with scroll snapping have native support for selecting items with the arrow keys, so this functionality should be built into the browsers/we should not need to rebuild it ourselves.

@liamdebeasi the popover implementation for keyboard interactions is preventing the arrow navigation.

https://github.com/ionic-team/ionic-framework/blob/next/core/src/components/popover/popover.tsx#L388

Specifically the `preventDefault()` - https://github.com/ionic-team/ionic-framework/blob/55db38ddc541c2632c7d3e4e4c9400ff5b5dfe8c/core/src/components/popover/utils.ts#L375

Do you recall the intention behind that? Looks like up/down/home are prevented, but left and right are not.

https://github.com/ionic-team/ionic-framework/blob/55db38ddc541c2632c7d3e4e4c9400ff5b5dfe8c/core/src/components/popover/utils.ts#L420

File: core/src/components/popover/popover.tsx [hunk 1]
```diff
@@ -208,6 +208,15 @@ export class Popover implements ComponentInterface, PopoverInterface {
...
   @Prop() isOpen = false;
 
+  /**
+   * If `true`, the popover will cancel browser keyboard event bindings to prevent
+   * scroll behavior in a popover using a list of items.
+   *
+   * Disable this behavior to allow the contents of the popover to handle their own
+   * keyboard interactions (i.e. when using CSS scroll snap).
+   */
+  @Prop() cancelBrowserKeyboardEvents = true;
```
Comment (liamdebeasi) [inline core/src/components/popover/popover.tsx:218]: What do you think about making this a prop that disables the keyboard events when `true`? I think if developers want the browser keyboard events to be handled normally, it's likely they don't want Ionic getting in the way either. I can't think of a use case where we would want both the browser and Ionic to handle the keyboard events.

Comment (sean-perkins) [reply core/src/components/popover/popover.tsx]: Makes sense to me! Verified locally that all the keyboard navigation built-ins work the same after disabling the popover's keyboard event handlers.

I went with `disableKeyboardEvents` for the property, open to suggestions. We don't appear to have a definite pattern for enable vs. disable language for properties.

Comment (liamdebeasi) [reply core/src/components/popover/popover.tsx]: I try to name public APIs using positive language, that way it's easier to reason about when negating values. For example `!popoverEl.enableKeyboardEvent` is easier to reason about than `!popoverEl.disableKeyboardEvent`

File: core/src/components/popover/popover.tsx [hunk 2]
```diff
@@ -208,6 +208,17 @@ export class Popover implements ComponentInterface, PopoverInterface {
    */
   @Prop() isOpen = false;
 
+  /**
+   * If `true` the popover will not register it's own keyboard event handlers.
```
Comment (liamdebeasi) [inline core/src/components/popover/popover.tsx:212]:
```suggestion
   * If `true` the popover will not register its own keyboard event handlers.
```
```suggestion
   * If `true` the popover will not register it's own keyboard event handlers.
```

File: core/src/components/popover/popover.tsx [hunk 3]
```diff
@@ -208,6 +208,17 @@ export class Popover implements ComponentInterface, PopoverInterface {
...
+   * This allows the contents of the popover to handle their own keyboard interactions.
+   *
+   * If `false`, the popover will register it's own keyboard event handlers for
```
Comment (liamdebeasi) [inline core/src/components/popover/popover.tsx:215]:
```suggestion
   * If `false`, the popover will register its own keyboard event handlers for
```
```suggestion
   * If `false`, the popover will register it's own keyboard event handlers for
```

**Review by liamdebeasi — APPROVED** [review]
The code looks good minus a few comment typos. Thoughts on making this a private property for now? I know we discussed about some hypothetical use cases developers may run into, but I think I would prefer to wait for devs to come to us with their use cases before we commit to having this be part of the public API.

**Comment by appy-one** [comment]
I've just created a new issue about keyboard navigation in a textarea in popovers (#24633), I though I'd share this here because it is affected by the same `preventDefault`s in overlays.ts

**Pull request merged**